cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionService.java

```diff
@@ -0,0 +1,205 @@
+      LOG.warn("Timeout in waiting for Spark program to stop: {}", programRunId);
+    }
+    httpServer.stopAndWait();
+  }
+
+  public void shutdownNow() {
+    stopLatch.countDown();
+    stop();
+  }
+
+  /**
+   * The {@link HttpHandler} for communicating with the Spark driver.
+   */
+  public final class SparkControllerHandler extends AbstractHttpHandler {
+
+    /**
+     * Handles heartbeat request from the running Spark program.
+     */
+    @POST
+    @Path("/v1/spark/{programName}/runs/{runId}/heartbeat")
+    public synchronized void heartbeat(HttpRequest request, HttpResponder responder,
+                                       @PathParam("programName") String programName,
+                                       @PathParam("runId") String runId) throws Exception {
+      if (stopLatch.await(0, TimeUnit.SECONDS)) {
+        throw new BadRequestException(
+          String.format("Spark program '%s' is already stopped. Heartbeat is not accepted.", programRunId));
+      }
+
+      validateRequest(programName, runId);
+      updateWorkflowToken(request.getContent());
+
+      // If the stop was requested, send the "stop" command
+      if (stopping.get()) {
+        Command.Builder.of("stop");
+        responder.sendJson(HttpResponseStatus.OK, SparkCommand.STOP);
+      } else {
+        responder.sendStatus(HttpResponseStatus.OK);
+      }
+    }
+
+    /**
+     * Handles execution completion request from the running Spark program.
+     */
+    @PUT
+    @Path("/v1/spark/{programName}/runs/{runId}/completed")
+    public synchronized void completed(HttpRequest request, HttpResponder responder,
+                                       @PathParam("programName") String programName,
+                                       @PathParam("runId") String runId) throws Exception {
+      validateRequest(programName, runId);
+      try {
+        updateWorkflowToken(request.getContent());
+        responder.sendStatus(HttpResponseStatus.OK);
+      } finally {
+        stopLatch.countDown();
+      }
+    }
+
+    /**
+     * Verifies the call is from the right client.
+     */
+    private void validateRequest(String programName, String runId) throws Exception {
+      if (!programRunId.getProgram().equals(programName)) {
+        throw new BadRequestException(
+          String.format("Request program name '%s' is not the same as the context program name '%s",
+                        programName, programRunId.getProgram())
+        );
+      }
+      if (runId == null || !programRunId.getRun().equals(runId)) {
+        throw new BadRequestException(
+          String.format("Request runId '%s' is not the same as the context runId '%s'", runId, programRunId.getRun())
+        );
+      }
+    }
+
+    /**
+     * Updates {@link WorkflowToken} of the program. It is a json Map<String, String> in the request body.
+     */
+    private void updateWorkflowToken(ChannelBuffer requestBody) {
+      if (!requestBody.readable()) {
+        return;
```

Comment:
Why? It is part of the protocol and is expected. See https://issues.cask.co/browse/CDAP-3008 for the protocol